staging/src/k8s.io/cli-runtime/pkg/resource/builder.go

```diff
@@ -1187,6 +1195,17 @@ func HasNames(args []string) (bool, error) {
 	return hasCombinedTypes || len(args) > 1, nil
 }
 
+// expandIfFilePattern returns all the filenames that match the input pattern
+// or the filename if it is a specific filename and not a pattern.
+// If the input is a pattern and it yields no result it will result in an error.
+func expandIfFilePattern(pattern string) ([]string, error) {
+	matches, err := filepath.Glob(pattern)
+	if err == nil && len(matches) == 0 {
+		return nil, fmt.Errorf("pattern did not yield any results")
```

Comment:
My shell doesn't expand when the pattern doesn't match any result. It goes from "do what you mean" to "do what you said", which is not unreasonable.

e.g.:
```
$ mkdir test && 
   cd test &&
   >test.yaml &&
   echo *.yaml *.json
test.yaml *.json
```

Reply:
On the other hand, Git probably does something similar to what you're proposing here:
```
$ git init test &&
   cd test &&
   >test.yaml &&
   git add '*.yaml' '*.json'
Initialized empty Git repository in /test/.git/
fatal: pathspec '*.json' did not match any files
```

So they're both probably fine.

Reply:
If both are fine but using the original string that yielded no matches resolves your other concern about breaking changes then I'm open to making that adjustment. i.e. someone decided to name their file `my?file.yaml` then `kubectl apply -f 'my?file.yaml'` would continue to work but would provide no feedback to somebody else actually wanting to expand the pattern `my?file.yaml` when it has no matches.

Reply:
The code as written will yield the following error:
```
file pattern *.json is not valid: pattern did not yield any results
```
which isn't great.

Can we do that instead?
```
pattern "*.json" is not valid: no match
```